caas/kubernetes/provider/resources/applier.go

```diff
@@ -52,29 +56,73 @@ func (op *operation) process(ctx context.Context, rollback Applier) error {
 	} else if err != nil {
 		return errors.Annotatef(err, "checking if resource %q exists or not", existingRes.ID())
 	}
-	if found {
-		ver := op.resource.GetObjectMeta().GetResourceVersion()
-		if ver != "" && ver != existingRes.GetObjectMeta().GetResourceVersion() {
-			id := op.resource.ID()
-			return errors.Annotatef(errConflict, "%s %s", id.Type, id.Name)
-		}
-	}
+
 	switch op.opType {
 	case opApply:
-		err = op.resource.Apply(ctx)
-		if found {
-			// apply the previously existing resource.
-			rollback.Apply(existingRes)
-		} else {
-			// delete the new resource just created.
-			rollback.Delete(op.resource)
+		err = retry.Call(retry.CallArgs{
+			Func: func() error {
+				err := op.resource.Apply(ctx)
+				if errors.Is(err, errConflict) {
+					_ = op.resource.Get(ctx) // refresh resource version
+					return err
+				}
+				// this might happen if it were deleted by another apiserver.
+				if k8serrors.IsNotFound(err) {
+					return err
+				}
+				return errors.Annotatef(err, "applying resource %q", op.resource.ID().Name)
+			},
+			IsFatalError: func(err error) bool {
+				return !k8serrors.IsConflict(err) && !errors.Is(err, errConflict)
+			},
+			Clock:       jujuclock.WallClock,
+			Attempts:    5,
+			Delay:       time.Second,
+			BackoffFunc: retry.ExpBackoff(time.Second, 5*time.Second, 1.5, true),
+		})
+
+		// we do not want to rollback if the resource is not found, not applied or
+		// it has been deleted by another apiserver during the retry call process.
+		if err == nil {
+			if found {
+				// apply the previously existing resource.
+				rollback.Apply(existingRes)
```

Comment:
We this always work as expected? If something else has modified the resource we are not guaranteed that existingRes will be the thing that we end up with I don't think. We'd end up patching the current version of the resource that exists at the time.

Reply:
hm not sure if that has always worked as expected. I agree that this may happen though, I was also thinking if it might be possible for two separate controllers in the same cluster to apply on a resource. Should we rely on conflict-retry + idempotent delete and remove the rollback feat for apply for now? 

These are what I'm thinking of for a temporary fix:
1. Retries on IsConflict and continue to treat NotFound as success for deletes.
2. We rollback/cleanup up only the newly created, and leave existing objects alone